llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -901,7 +901,7 @@ void WaitcntBrackets::updateByEvent(const SIInstrInfo *TII,
     }
   } else /* LGKM_CNT || EXP_CNT || VS_CNT || NUM_INST_CNTS */ {
     // Match the score to the destination registers.
-    for (unsigned I = 0, E = Inst.getNumOperands(); I != E; ++I) {
+    for (unsigned I = 0, E = Inst.getNumExplicitOperands(); I != E; ++I) {
```

Comment:
I can never remember how all of these counts work with the 2 notions of "implicit". Does this catch the implicit-def present in the instruction definition? e.g. implicit-def $vcc on v_cmp_*

Reply:
This should also get a comment

Reply:
> I can never remember how all of these counts work with the 2 notions of "implicit". Does this catch the implicit-def present in the instruction definition? e.g. implicit-def $vcc on v_cmp_*

For non-variadic instructions it returns a getNumOperands() from its MCInstrDesc. I.e. it shall include implicit defs and uses from the instruction description. Besides, there is no way to load vcc, so there is no need to wait for such load. I have also checked there is no way to load M0, a common real implicit use, as scalar loads use XM0 class for sdst. I cannot think of any other implicit operands which may matter.

Reply:
Hm... Actually it excludes implicit uses and defs even from the instruction description. But I still cannot see any relevant registers which may need a wait.

Reply:
If you want I can add to getNumExplicitArguments() also getDesc().NumImplicitDefs and getDesc().NumImplicitUses, but I do not see a case for it. I am also not sure the order of implicit operands will be maintained.

Reply:
> Besides, there is no way to load vcc

Yes there is! You can load it with s_load or s_buffer_load.

Reply:
> > Besides, there is no way to load vcc
> 
> Yes there is! You can load it with s_load or s_buffer_load.

*** Bad machine code: Illegal physical register for instruction ***
- function:    load_vcc_wait
- basic block: %bb.0  (0x87a5980)
- instruction: $vcc = S_LOAD_DWORD_IMM $sgpr10_sgpr11, 0, 0
- operand 0:   $vcc
$vcc is not a SReg_32_XM0_XEXEC register.
LLVM ERROR: Found 2 machine code errors.

Reply:
> > Besides, there is no way to load vcc
> 
> Yes there is! You can load it with s_load or s_buffer_load.

Ouch.

Reply:
OK, I will add this test and see where the rabbit hole goes...

Reply:
I have added the test with vcc_lo load and implicit use of vcc, it has the wait as expected.

The reason it is correct because only memory instructions arrive to the modified loop, VALU is processed in a different place. Memory instructions may implicitly use m0 and flat_scr, but these are not loadable.